templates/extras/simp.md

```diff
@@ -1,111 +1,254 @@
+goals or hypotheses, by using repeated rewrites. Here is an example
+(using `mathlib`).
 
 ```lean
-namespace hidden
+import algebra.group.defs
 
-definition cong (a b m : ℤ) : Prop := ∃ n : ℤ, b - a = m * n
+variables (G : Type) [group G] (a b c : G)
 
-notation a ` ≡ ` b ` mod ` m  := cong a b m
-set_option trace.simplify true
-theorem cong_refl (m : ℤ) : ∀ a : ℤ, a ≡ a mod m :=
+example : a * a⁻¹ * 1 * b = b * c * c⁻¹ :=
 begin
-intro a,
-unfold cong,
-existsi (0:ℤ),
-simp
+  simp
 end
-
-end hidden
 ```
 
-If you do this exercise you will discover firstly that `simp` spends a lot of its time trying random
-lemmas and then giving up very shortly afterwards, and also that the `unfold` command is also
-underlined in green -- Lean seems to apply `simp` when you do an `unfold` as well (apparently
-`unfold` just asks `simp` to do its dirty work for it -- `unfold X` is close to `simp only [X]`).
-
-If you only want to see what worked rather than all the things that didn't, you could try
-`set_option trace.simplify.rewrite true`.
+How would a human solve that goal? They would notice that `a * a⁻¹ =
+1`, that `1 * 1 = 1` and so on, until they had simplified the example
+to `b = b`, which is obviously true. This is also what the simplifier
+is doing. Indeed, if you add `set_option trace.simplify.rewrite true`
+above the example, then a squiggly blue underline will appear under
+`simp` (in VS Code) and clicking on this will show you the sequence of
+rewrites that `simp` discovered.
 
 ## Simp lemmas
 
-In case you want to train `simp` to use certain extra lemmas (for example because they're coming up
-again and again in your work) you can add new lemmas for yourself. For example in mathlib in
-`algebra/ring.lean` we find the line
+So how did Lean's simplifier know that `a * a⁻¹ = 1`? It is because
+the corresponding lemma in `algebra.group.defs` is tagged with the
+`simp` attribute:
+
+```lean
+@[simp] lemma mul_right_inv (a : G) : a * a⁻¹ = 1 := ...
+```
+
+This `@[simp]` at the beginning just "tags" `mul_right_inv` with the
+`simp` attribute. We call lemmas tagged in this way "`simp`
+lemmas". Here are some examples of `simp` lemmas in mathlib:
 
 ```lean
-@[simp] theorem ne_zero (u : units α) : (u : α) ≠ 0
+@[simp] theorem nat.dvd_one {n : ℕ} : n ∣ 1 ↔ n = 1 := ...
+@[simp] theorem mul_eq_zero {a b : ℕ} : a * b = 0 ↔ a = 0 ∨ b = 0 := ...
+@[simp] theorem list.mem_singleton {a b : α} : a ∈ [b] ↔ a = b := ...
+@[simp] theorem set.set_of_false : {a : α | false} = ∅ := ...
 ```
 
-This lemma is then added to `simp`'s armoury. Note several things however.
+When the simplifier is attempting to simplify a term `T` it looks
+through the `simp` lemmas known to the system at that time, in a
+sensible way (there are over ten thousand of them currently in
+`mathlib`!), and if it runs into a lemma of the form `A = B` or `A ↔
+B`, for which `A` shows up in `T`, it rewrites the lemma at `T` to
+replace it with a `B`, and starts again.
```

Comment:
Worth mentioning here (with an example, maybe) that `simp` works "inside-out".

Reply:
I don't know what this means. Can you sketch the idea? I still have no idea how simp works, I just now know what it does.

Reply:
[Here's](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Simp.20normal.20form/near/188829903) a post by Gabriel that I starred in Zulip about this:

> The simplifier simplifies terms inside-out, that is, it first simplifies the arguments of a function before simplifying the function. Hence a simp lemma is only ever used to rewrite terms where the arguments are already fully simplified (= in simp-normal form). Corollary: if the left-hand side of your simp lemma has an argument that is not fully simplified, then this simp lemma is never used to rewrite a term by the simplifier and is hence useless.